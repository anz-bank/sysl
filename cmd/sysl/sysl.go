package main

import (
	"fmt"
	"os"
	"strings"

	"github.com/anz-bank/sysl/pkg/arrai"

	"github.com/anz-bank/sysl/pkg/cfg"
	"github.com/anz-bank/sysl/pkg/syslutil"
	"github.com/pkg/errors"
	"github.com/sirupsen/logrus"
	"github.com/spf13/afero"
	"gopkg.in/alecthomas/kingpin.v2"
)

const syslRootMarker = ".sysl"

// main3 is the real main function. It takes its output streams and command-line
// arguments as parameters to support testability.
func main3(args []string, fs afero.Fs, logger *logrus.Logger) error {
	flags, err := syslutil.PopulateCMDFlagsFromFile(args)
	if err == nil && len(flags) > 0 {
		// apply flags in file
		args = flags
	}

	syslCmd := kingpin.New("sysl", "System Specification Language Toolkit")
	syslCmd.Version(fmt.Sprintf("sysl %s %s", Version, BuildOS))
	syslCmd.UsageTemplate(kingpin.SeparateOptionalFlagsUsageTemplate)

	(&debugTypeData{}).add(syslCmd)

	runner := cmdRunner{}
	if err := runner.Configure(syslCmd); err != nil {
		return err
	}

	selectedCommand, err := syslCmd.Parse(args[1:])
	if err != nil {
		return err
	}

	return runner.Run(selectedCommand, fs, logger)
}

type debugTypeData struct {
	loglevel string
	verbose  bool
}

//nolint:unparam
func (d *debugTypeData) do(_ *kingpin.ParseContext) error {
	if d.verbose {
		d.loglevel = cfg.LogLevel
	}
	// Default info
	syslutil.SetLogLevel(d.loglevel)

	logrus.Debugf("Logging: %+v", *d)
	return nil
}
func (d *debugTypeData) add(app *kingpin.Application) {
	var levels []string
	for l := range syslutil.LogLevels {
		if l != "" {
			levels = append(levels, l)
		}
	}
	app.Flag("log", fmt.Sprintf("log level: [%s]", strings.Join(levels, ","))).
		HintOptions(levels...).
		Default("warn").
		StringVar(&d.loglevel)
	app.Flag("verbose", "enable verbose logging").Short('v').BoolVar(&d.verbose)
	app.PreAction(d.do)
}

// main2 calls main3 and handles any errors it returns. It takes its output
// streams and command-line arguments and even main3 as parameters to support
// testability.
func main2(
	args []string,
	fs afero.Fs,
	logger *logrus.Logger,
	main3 func(args []string, fs afero.Fs, logger *logrus.Logger) error,
) int {
	if err := main3(args, fs, logger); err != nil {
		arraiErr, ok := errors.Cause(err).(arrai.ExecutionError)
		if ok {
			logger.Debugln(err)
			logger.Errorln(arraiErr.ShortMsg)
		} else {
			var exitCode = 1
			if err, ok := err.(syslutil.Exit); ok {
				exitCode = err.Code
			}

			switch exitCode {
			case 0:
				logger.Infoln(err.Error())
			case 2:
				logger.Warnln(err.Error())
			default:
				logger.Errorln(err.Error())
			}

			return exitCode
		}
	}
	return 0
}

// main is as small as possible to minimise its no-coverage footprint.
func main() {
	if rc := main2(os.Args, afero.NewOsFs(), logrus.StandardLogger(), main3); rc != 0 {
		os.Exit(rc)
	}
}
