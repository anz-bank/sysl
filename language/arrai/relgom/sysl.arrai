(
    baseType: \model //.fn.fix(\baseType
        \type
        (
            let path = type("typeRef")("ref")("path").a;
            baseType(model("types")(path(0).s)("relation")("attrDefs")(path(1).s))
        ) if "typeRef" <: (type => .@) else type
    ),
)
