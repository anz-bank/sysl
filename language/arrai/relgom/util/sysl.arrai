let attrs = \node
    node('attrs') if 'attrs' <: (node => .@);

let patterns = \node
    let attrs = attrs(node);
    (attrs('patterns')('a')('elt').a => .@item('s').s) if 'patterns' <: (attrs => .@);

let isAutoinc  = \attrDef "autoinc" <: patterns(attrDef);
let isPK       = \attrDef "pk" <: patterns(attrDef);
let isOpt      = \attrDef "opt" <: (attrDef => .@) && attrDef("opt");
let isRequired = \attrDef (isPK(attrDef) || !isOpt(attrDef)) && !isAutoinc(attrDef);

(
    :attrs,
    :isAutoinc,
    :isPK,
    :isOpt,
    :isRequired,
    :patterns,

    baseType: \model //.fn.fix(\baseType
        \type
            (
                let path = type('typeRef')('ref')('path').a;
                baseType(model('types')(path(0).s)('relation')('attrDefs')(path(1).s))
            ) if 'typeRef' <: (type => .@) else type
    ),

    fkey: \type type('typeRef') if 'typeRef' <: (type => .@),

    namedTypesInSourceOrder: \types
        types => (name: .@, type: .@value) orderby [.type('sourceContext')('start')('line')],
)
