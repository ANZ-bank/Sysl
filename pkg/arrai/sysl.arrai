# Performs transformations of Sysl models, producing "better" Sysl models.

# Loads and returns a Sysl model.
let load = \path
    let syslDescriptor = //encoding.proto.descriptor(//os.file('sysl.pb'));
    //encoding.proto.decode(syslDescriptor, 'Module' , //os.file(path));

# Extracts the set of pattern attributes of an element.
let patterns = \x x.attrs?('patterns')?.a.elt:{} => .@item.s;

# A macro to convert `ret` statement payload string to structured tuples.
# TODO: handle dynamic type (int, float etc)
let payloadMacro = (
    @grammar: {://grammar.lang.wbnf:
        payload -> (status ("<:" type)? | (status "<:")? type) attr?;
        type -> [^\[\n]*;
        status -> ("ok"|"error"|[1-5][0-9][0-9]);
        attr -> %!Array(nvp|modifier);
        nvp_item -> ('"' ([^"\\] | [\\][\\brntu'"])* '"' | "'" ([^''])* "'") | array=%!Array(nvp_item) | dict=%!Dict(nvp_item);
        nvp ->  name=\w+ "=" nvp_item;
        modifier -> "~" name=[\w\+]+;
        .wrapRE -> /{\s*()\s*};
        .macro Array(child) {
            "[" (child):"," "]"
        }
        .macro Dict(child) {
            "{" entry=(key=child ":" value=child):"," "}"
        }
    :},
    @transform: (payload: //{./util}.simpleTransform),
);

# TODO: this is assuming that attribute values are similar to arrai strings
let rec buildNvp = \nvp cond nvp {
    (array: (nvp_item: i, ...), ...): (a: i => (:.@, @item: buildNvp(.@item))),
    (dict: (entry: i, ...), ...): (d: i => (
        @     : buildNvp(.@item.key.nvp_item),
        @value: buildNvp(.@item.value.nvp_item)
    )),
    _: //eval.value(//seq.join('', nvp.''))
};

let name = //{./util}.remOffset;

# tries to fix type without full reference
let fixType = \apps \appName \typePayload
    # add trim whitespace function
    let typePayload = //{./util}.trimWhitespace(typePayload);
    let (:app, :type, :field) = //{./util}.unpackType(typePayload);
    let pack = //{./util}.packType;
    cond app {
        (''): pack((app: appName, :type, :field)),
        _: cond {
            (app <: apps): typePayload,
            _: pack((app: appName, type: [app] ++ type, :field))
        }
    }
;
# Converts a return statement payload into a tuple of its parts.
let parseReturn = \payload \apps \currApp
    let ret = //{./util}.invokeMacro(payloadMacro, payload);
    let retType = ret.type?:'';
    ret +> (
        type: fixType(apps, currApp, retType),
        attr: //grammar -> .parse(payloadMacro.@grammar, 'payload', payload).attr?:() -> \attr
                (
                    nvp: attr.nvp?:{} => (@: name(.@item.name.''), @value: buildNvp(.@item.nvp_item)),
                    modifier: attr.modifier?:{} => name(.@item.name.'')
                )
    )
;

# Replaces all `ret` payloads with structured tuples.
let polishReturns = \m
    let apps = m.apps => .@;
    let tweak = \. \currApp
        . +> (ret: .ret +> (payload: parseReturn(.ret.payload, apps, currApp)));

    m +> (apps: m.apps?:{} >>> \currApp \a
        a +> (endpoints: a.endpoints?:{} >> \p
            p +> (stmt: p.stmt?:[] >> \s
                cond s.ret?:{} {{}: s, _: tweak(s, currApp)}
            )
        )
    )
;

# Applies polish functions to the appropriate elements of a Sysl model.
let polish = \m
    polishReturns(m)
    # TODO: More polish.
;

# Returns a normalised, relational model of a Sysl model.
let normalize = \m
    let apps = m.apps => \(@:appName, @value:app)
        (
            :appName,
            appPatterns: patterns(app),
        )
    ;
    let endpoints = (m.apps => \(@:appName, @value:app)
        app.endpoints?:{} => \(@:epName, ...)
            (
                :appName,
                :epName,
            )
    ) -> //rel.union(.);
    let stmts = (m.apps => \(@:appName, @value:app)
        (app.endpoints?:{} => \(@:epName, @value:ep)
            ep.stmt?:{} =>
                (
                    :appName,
                    :epName,
                    stmt: .@item +> (
                        # TODO: Use stmt.source_context once populated by parser.
                        # https://github.com/anz-bank/sysl/issues/990
                        source_context: ep.source_context +> (:.@),
                    ),
                )
        ) -> //rel.union(.)
    ) -> //rel.union(.);
    let calls = stmts where .stmt.call?:{} => \(:appName, :epName, :stmt)
        (
            from: (
                :appName,
                :epName,
            ),
            to: (
                appName: stmt.call.target.part.@item,
                epName: stmt.call.endpoint,
            ),
            :stmt.source_context,
        )
    ;
    let types = (m.apps => \(@:appName, @value:app)
        app.types?:{} => \(@:typeName, @value: t)
            (
                :appName,
                :typeName,
            )
        ) -> //rel.union(.) where .
    ;
    let fields = (m.apps => \(@:appName, @value:app)
        (app.types?:{} => \(@:typeName, @value:type)
            type.relation?.attr_defs?:{} => \(@:fieldName, @value:field)
                let patterns = patterns(field);
                (
                    :appName,
                    :typeName,
                    :fieldName,
                    fieldType: cond field {
                        (primitive:p, ...): p,
                        (type_ref:(ref:(path:p), ...), ...): //seq.join('.', p),
                        _: '?',
                    },
                    fk: cond field.type_ref?.ref.path:{} {
                        [t, f]: (:appName, typeName: t, fieldName: f)
                    },
                    :patterns,
                    pk: !!({'pk'} & patterns),
                    opt: field.opt?:{},
                )
        ) -> //rel.union(.) where .
    ) -> //rel.union(.);
    let transforms = //{./dfa/transforms}(m.apps);
    (
        :apps,
        :endpoints,
        :calls,
        :types,
        :fields,
        :transforms,
        # TODO:
        # :annotations, (aka attributes)
        :stmts,
        # :tags, (aka patterns)
    )
;

# Export the functions from the module.
(
    :load,
    :normalize,
    :parseReturn,
    :patterns,
    :polish,
)
#  -> //fmt.pretty(.polish(.load('out/model.pb')))
#  -> //fmt.pretty(parseReturn('ok <: Foo [dataflow={"a.b.c": "b", "c": ["d", "e"]}]'))
