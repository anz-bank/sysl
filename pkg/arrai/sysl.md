# Sysl

`sysl.arrai` contains helper functions to make working with Sysl models more convenient.

Examples in this file assume you've imported `sysl.arrai` using the following syntax:
```arrai
let sysl = //{github.com/anz-bank/sysl/pkg/arrai/sysl};
```

## polish()

`polish` takes an imported Sysl model and returns a copy with a cleaner internal structure, making it easier to work with. It applies the following tweaks:

- Parse `return` statement payloads into `(status, type)` tuples.

## ~~normalize()~~ *[Deprecated]*

`normalize` takes an imported Sysl model and returns an equivalent relational model. The relational model is a tuple structured like a database, with tables for each Sysl concept (e.g. `apps`, `endpoints`, `types`, `calls`, etc.) that can be naturally joined with the `<&>` operator.

The relational model is a powerful option for transformations that resemble queries with joins and filters (i.e. `where` clauses). See the diagram scripts for examples.

## load()

Loads a sysl model that is stored in a protobuf file. To generate such a file for `myModel.sysl`, run the Sysl tool with the following arguments:
```sh
sysl protobuf --output=myModel.pb --mode=pb myModel.sysl
```

You can then to load the model into arr.ai, use the following:
```arrai
let model = sysl.load('myModel.pb');
```

## loadBytes()

Loads a sysl model protobuf-encoded bytes (e.g. from `//os.stdin`, `//[//encoding.bytes]{./path}` or `//os.file()`). For example, you can generate those bytes from `myModel.sysl` and pipe them into an arr.ai script `myScript.arrai` using the following command:
```sh
sysl protobuf --mode=pb myModel.sysl | arrai r myScript.arrai
```

**myScript.arrai** could then contain the following:
```arrai
let model = sysl.loadBytes(//os.stdin);
```

## normalizeNew()

After a model was loaded with `load()` or `loadBytes()`, it can then be transformed to the relational model using `normalizeNew()` as such:

```arrai
var relModel = sysl.newNormalize(model);
```
