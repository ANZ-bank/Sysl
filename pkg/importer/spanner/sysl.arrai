# Transforms that generate sysl from arrai based sql model
# size returns the size of an attribute
let size = \length
    cond {
        length = 'MAX': '',
        //seq.has_prefix('0x', length): $`(${//seq.trim_prefix('0x', length)})`,
        length > 0: $`(${length})`,
    };

# hasAttributePattern checks whether an attribute has associated patterns
let hasAttributePattern = \entity \attr
    cond {
        attr.options?:{}: true,
        (entity.primary_key where //seq.contains(attr.name, .)): true,
        let fks = (entity.foreign_keys => ((.foreign_keys where .attribute = attr.name) =>
            .reference_attribute orderby .)) where .; fks: true,
        attr.length = 'MAX': true,
        //seq.has_prefix('0x', attr.length): true,
    };

# hasEntityPattern checks whether an entity has associated patterns
let hasEntityPattern = \entity \model
    entity.primary_key | entity.foreign_keys | entity.cluster | model.indexes;

# attributePatternDelimiter returns the token based on attribute value
let attributePatternDelimiter = \entity \attr \token
    cond {
        hasAttributePattern(entity, attr): token
    };

# entityPatternDelimiter returns the token based on entity value
let entityPatternDelimiter = \entity \model \token
    cond {
        hasEntityPattern(entity, model): token
    };

# sortingOrder determines and appends sorting order
let sortingOrder = \e \attr
    let keyOrder = //seq.split('(', //seq.join('', (e.primary_key where //seq.contains(attr.name, .)) orderby .));
    cond {
        keyOrder count > 1: cond keyOrder(1) {
            'asc)': ', ~asc',
            'desc)': ', ~desc'
        }
    };

# compareColumnOrder compares the col order between primary key and table
let compareColumnOrder = \entity
    let pk = (entity.primary_key orderby .) >> //seq.split('(', .) >> .(0);
    cond {
        pk & (entity.attributes >> .name): false,
        _: true,
    };

# attributePattern generates the pattern constituents depending on the mode
let attributePattern = \entity \attr \mode
    cond {
        mode = 0 && (entity.primary_key where //seq.contains(attr.name, .)): $`~pk${sortingOrder(entity, attr)}`,
        let fks = (entity.foreign_keys => ((.foreign_keys where .attribute = attr.name) => .reference_attribute orderby .)) where .;
        mode = 1 && fks: '~fk',
        mode = 2 && attr.options:
            let opt = //seq.split('=',attr.options);
            $`${opt(0)}='${opt(1)}'`,
        mode = 3 && attr.length = 'MAX': '~max',
        mode = 4 && //seq.has_prefix('0x', attr.length): '~hex',
    };

# entityPattern generates the pattern constituents depending on the mode
let entityPattern = \entity \model \mode
    cond {
        mode = 0 && entity.primary_key count > 1 && compareColumnOrder(entity):
            $`primary_key="${//seq.join(', ', entity.primary_key orderby .)}"`,
        mode = 1 && entity.cluster:
            //seq.join('', entity.cluster >> $`interleave_in_parent="${.interleaved_in}", interleave_on_delete="${.on_delete}"`),
        mode = 2 && entity.foreign_keys:
            $`foreign_keys=[${//seq.join(', ',(entity.foreign_keys => \keys $`["constraint:${keys.constraint_name}` ++
            $`","columns:${//seq.join(', ', (keys.foreign_keys => .attribute) orderby .)}"]`) orderby .)}]`,
        let relatedIndices = ((model.indexes orderby .) where .@item.table_name = entity.name); mode = 3 && relatedIndices:
            $`indexes=[${//seq.join(', ', ((model.indexes orderby .) where .@item.table_name = entity.name) >>
            $`"name:${.table_name}` ++
            $`","unique:${cond {(.unique = 'UNIQUE'):'true', _:'false'}}` ++
            $`","null_filtered:${cond {(.nullfiltered = 'NULL_FILTERED'):'true', _:'false'}}` ++
            $`","key_parts:${//seq.join(', ', .key_part orderby .)}` ++
            cond {.storing_col: $`","storing:' ${//seq.join(', ',.storing_col)}`}  ++
            cond {.interleaved_table: $`","interleave_in:' ${.interleaved_table}`})}"]`,
    };

# allEntityPatterns generates the pattern string for an entity
let allEntityPatterns = \entity \model
    $`${entityPatternDelimiter(entity, model, '[')}${cond {
        hasEntityPattern(entity, model):
        //seq.join(', ',[entityPattern(entity, model, 0), entityPattern(entity, model, 1),
        entityPattern(entity, model, 2), entityPattern(entity, model, 3)] where .@item != '')
    }}${entityPatternDelimiter(entity, model, ']')}`;

# allAttributePatterns generates the pattern string for an entity attribute
let allAttributePatterns = \entity \attr
    $`${attributePatternDelimiter(entity,attr,'[')}${cond {
        hasAttributePattern(entity,attr):
        //seq.join(', ',[attributePattern(entity,attr,2), attributePattern(entity,attr,0), attributePattern(entity,attr,1),
        attributePattern(entity,attr,3), attributePattern(entity,attr,4)] where .@item != '')
    }}${attributePatternDelimiter(entity,attr,']')}`;

# typeInfo generates the type info for an attribute
let typeInfo = \entity \attr \type \isArray
    let fkAttr = entity.foreign_keys => ((.foreign_keys where .attribute = attr.name) =>
        .reference_attribute orderby .) where .;
    let fkTable = entity.foreign_keys => ((.foreign_keys where .attribute = attr.name) =>
        .reference_table orderby .) where .;
    cond {
        fkAttr | fkTable:  //seq.join('', (fkTable => //seq.join('', .) orderby .)) ++ '.'
            ++ //seq.join('', (fkAttr => //seq.join('', .) orderby .)),
        isArray: $`sequence of ${type}`,
        _: type,
    };

# transformModel translates the empty model into sysl file
let transformModel = \model \package
    # sysl specification
    # https://github.com/anz-bank/sysl/blob/master/pkg/sysl/sysl.proto
    $`
        ########   THIS IS AUTOGENERATED BY sysl   ########
        ${model.schema => $`
            ${.name}${cond {package: $`[package="${package}"]`} }:
            ${//seq.sub('#>>>\n', '')($`
            ${(model.entities => \entity $`
            #>>>
                !table ${entity.name} ${allEntityPatterns(entity, model)}:
                    ${entity.attributes >>
                        $`
                            ${.name} <: ${typeInfo(entity, ., .type ++ size(.length), .array)}${cond {.nullable:'?'}} ${allAttributePatterns(entity, .)}` ::\i:
                        }`
                    )
                    orderby .::\i:
            }
            `)}
        ` orderby .:::}
    `;
(
    :transformModel,
)
