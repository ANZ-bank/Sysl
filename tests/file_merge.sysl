# file_merge.sysl
import file_merge_import.sysl

Foo :: Bar:
    !type User:
        Name <: string


    /user:
        DELETE:
            @action="Delete user"
            return ok <: string
        POST:
            @action="Add user"
            return error <: string

    !type Profile:
        Name <: string:
            @tag1 = "tag1"

    !type Profile:
        Desc <: string:
            @tag1 = "tag1"
