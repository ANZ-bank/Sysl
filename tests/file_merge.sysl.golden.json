{
 "apps": {
  "Foo :: Bar": {
   "name": {
    "part": [
     "Foo",
     "Bar"
    ]
   },
   "endpoints": {
    "DELETE /user": {
     "name": "DELETE /user",
     "attrs": {
      "action": {
       "s": "Delete user"
      },
      "patterns": {
       "a": {
        "elt": [
         {
          "s": "rest"
         }
        ]
       }
      }
     },
     "stmt": [
      {
       "ret": {
        "payload": "ok <: string"
       },
       "sourceContext": {
        "file": "dev/sysl/tests/file_merge.sysl",
        "start": {
         "line": 11,
         "col": 12
        },
        "end": {
         "line": 11,
         "col": 18
        }
       },
       "sourceContexts": [
        {
         "file": "dev/sysl/tests/file_merge.sysl",
         "start": {
          "line": 11,
          "col": 12
         },
         "end": {
          "line": 11,
          "col": 18
         }
        }
       ]
      }
     ],
     "restParams": {
      "method": "DELETE",
      "path": "/user"
     },
     "sourceContext": {
      "file": "dev/sysl/tests/file_merge.sysl",
      "start": {
       "line": 9,
       "col": 8
      },
      "end": {
       "line": 12,
       "col": 8
      }
     },
     "sourceContexts": [
      {
       "file": "dev/sysl/tests/file_merge.sysl",
       "start": {
        "line": 9,
        "col": 8
       },
       "end": {
        "line": 12,
        "col": 8
       }
      }
     ]
    },
    "GET /user": {
     "name": "GET /user",
     "attrs": {
      "action": {
       "s": "Get user"
      },
      "patterns": {
       "a": {
        "elt": [
         {
          "s": "rest"
         }
        ]
       }
      }
     },
     "stmt": [
      {
       "ret": {
        "payload": "ok <: string"
       },
       "sourceContext": {
        "file": "dev/sysl/tests/file_merge_import.sysl",
        "start": {
         "line": 11,
         "col": 12
        },
        "end": {
         "line": 11,
         "col": 18
        }
       },
       "sourceContexts": [
        {
         "file": "dev/sysl/tests/file_merge_import.sysl",
         "start": {
          "line": 11,
          "col": 12
         },
         "end": {
          "line": 11,
          "col": 18
         }
        }
       ]
      }
     ],
     "restParams": {
      "method": "GET",
      "path": "/user"
     },
     "sourceContext": {
      "file": "dev/sysl/tests/file_merge_import.sysl",
      "start": {
       "line": 9,
       "col": 8
      },
      "end": {
       "line": 13,
       "col": 4
      }
     },
     "sourceContexts": [
      {
       "file": "dev/sysl/tests/file_merge_import.sysl",
       "start": {
        "line": 9,
        "col": 8
       },
       "end": {
        "line": 13,
        "col": 4
       }
      }
     ]
    },
    "POST /user": {
     "name": "POST /user",
     "attrs": {
      "action": {
       "s": "Add user"
      },
      "patterns": {
       "a": {
        "elt": [
         {
          "s": "rest"
         },
         {
          "s": "rest"
         }
        ]
       }
      }
     },
     "stmt": [
      {
       "ret": {
        "payload": "error <: string"
       },
       "sourceContext": {
        "file": "dev/sysl/tests/file_merge.sysl",
        "start": {
         "line": 14,
         "col": 12
        },
        "end": {
         "line": 14,
         "col": 18
        }
       },
       "sourceContexts": [
        {
         "file": "dev/sysl/tests/file_merge.sysl",
         "start": {
          "line": 14,
          "col": 12
         },
         "end": {
          "line": 14,
          "col": 18
         }
        }
       ]
      },
      {
       "ret": {
        "payload": "ok <: string"
       },
       "sourceContext": {
        "file": "dev/sysl/tests/file_merge_import.sysl",
        "start": {
         "line": 8,
         "col": 12
        },
        "end": {
         "line": 8,
         "col": 18
        }
       },
       "sourceContexts": [
        {
         "file": "dev/sysl/tests/file_merge_import.sysl",
         "start": {
          "line": 8,
          "col": 12
         },
         "end": {
          "line": 8,
          "col": 18
         }
        }
       ]
      }
     ],
     "restParams": {
      "method": "POST",
      "path": "/user"
     },
     "sourceContext": {
      "file": "dev/sysl/tests/file_merge_import.sysl",
      "start": {
       "line": 6,
       "col": 8
      },
      "end": {
       "line": 9,
       "col": 8
      }
     },
     "sourceContexts": [
      {
       "file": "dev/sysl/tests/file_merge.sysl",
       "start": {
        "line": 12,
        "col": 8
       },
       "end": {
        "line": 16,
        "col": 4
       }
      },
      {
       "file": "dev/sysl/tests/file_merge_import.sysl",
       "start": {
        "line": 6,
        "col": 8
       },
       "end": {
        "line": 9,
        "col": 8
       }
      }
     ]
    }
   },
   "types": {
    "Profile": {
     "tuple": {
      "attrDefs": {
       "Desc": {
        "primitive": "STRING",
        "attrs": {
         "tag1": {
          "s": "tag1"
         }
        },
        "sourceContext": {
         "file": "dev/sysl/tests/file_merge.sysl",
         "start": {
          "line": 21,
          "col": 8
         },
         "end": {
          "line": 23
         }
        },
        "sourceContexts": [
         {
          "file": "dev/sysl/tests/file_merge.sysl",
          "start": {
           "line": 21,
           "col": 8
          },
          "end": {
           "line": 23
          }
         }
        ]
       },
       "ID": {
        "primitive": "STRING",
        "sourceContext": {
         "file": "dev/sysl/tests/file_merge_import.sysl",
         "start": {
          "line": 14,
          "col": 8
         },
         "end": {
          "line": 14,
          "col": 14
         }
        },
        "sourceContexts": [
         {
          "file": "dev/sysl/tests/file_merge_import.sysl",
          "start": {
           "line": 14,
           "col": 8
          },
          "end": {
           "line": 14,
           "col": 14
          }
         }
        ]
       },
       "Name": {
        "primitive": "STRING",
        "attrs": {
         "tag1": {
          "s": "tag1"
         },
         "tag2": {
          "s": "tag2"
         }
        },
        "sourceContext": {
         "file": "dev/sysl/tests/file_merge_import.sysl",
         "start": {
          "line": 15,
          "col": 8
         },
         "end": {
          "line": 17
         }
        },
        "sourceContexts": [
         {
          "file": "dev/sysl/tests/file_merge.sysl",
          "start": {
           "line": 17,
           "col": 8
          },
          "end": {
           "line": 20,
           "col": 4
          }
         },
         {
          "file": "dev/sysl/tests/file_merge_import.sysl",
          "start": {
           "line": 15,
           "col": 8
          },
          "end": {
           "line": 17
          }
         }
        ]
       }
      }
     },
     "sourceContext": {
      "file": "dev/sysl/tests/file_merge_import.sysl",
      "start": {
       "line": 13,
       "col": 4
      },
      "end": {
       "line": 17
      }
     },
     "sourceContexts": [
      {
       "file": "dev/sysl/tests/file_merge.sysl",
       "start": {
        "line": 16,
        "col": 4
       },
       "end": {
        "line": 20,
        "col": 4
       }
      },
      {
       "file": "dev/sysl/tests/file_merge.sysl",
       "start": {
        "line": 20,
        "col": 4
       },
       "end": {
        "line": 23
       }
      },
      {
       "file": "dev/sysl/tests/file_merge_import.sysl",
       "start": {
        "line": 13,
        "col": 4
       },
       "end": {
        "line": 17
       }
      }
     ]
    },
    "User": {
     "tuple": {
      "attrDefs": {
       "Id": {
        "primitive": "STRING",
        "sourceContext": {
         "file": "dev/sysl/tests/file_merge_import.sysl",
         "start": {
          "line": 3,
          "col": 8
         },
         "end": {
          "line": 3,
          "col": 14
         }
        },
        "sourceContexts": [
         {
          "file": "dev/sysl/tests/file_merge_import.sysl",
          "start": {
           "line": 3,
           "col": 8
          },
          "end": {
           "line": 3,
           "col": 14
          }
         }
        ]
       },
       "Name": {
        "primitive": "STRING",
        "sourceContext": {
         "file": "dev/sysl/tests/file_merge.sysl",
         "start": {
          "line": 5,
          "col": 8
         },
         "end": {
          "line": 5,
          "col": 16
         }
        },
        "sourceContexts": [
         {
          "file": "dev/sysl/tests/file_merge.sysl",
          "start": {
           "line": 5,
           "col": 8
          },
          "end": {
           "line": 5,
           "col": 16
          }
         }
        ]
       }
      }
     },
     "sourceContext": {
      "file": "dev/sysl/tests/file_merge_import.sysl",
      "start": {
       "line": 2,
       "col": 4
      },
      "end": {
       "line": 3,
       "col": 14
      }
     },
     "sourceContexts": [
      {
       "file": "dev/sysl/tests/file_merge.sysl",
       "start": {
        "line": 4,
        "col": 4
       },
       "end": {
        "line": 5,
        "col": 16
       }
      },
      {
       "file": "dev/sysl/tests/file_merge_import.sysl",
       "start": {
        "line": 2,
        "col": 4
       },
       "end": {
        "line": 3,
        "col": 14
       }
      }
     ]
    }
   },
   "sourceContext": {
    "file": "dev/sysl/tests/file_merge_import.sysl",
    "start": {
     "line": 1
    },
    "end": {
     "line": 17
    }
   },
   "sourceContexts": [
    {
     "file": "dev/sysl/tests/file_merge.sysl",
     "start": {
      "line": 3
     },
     "end": {
      "line": 23
     }
    },
    {
     "file": "dev/sysl/tests/file_merge_import.sysl",
     "start": {
      "line": 1
     },
     "end": {
      "line": 17
     }
    }
   ]
  }
 }
}